File: Problems/Algorithms/40. Combination Sum II/CombinationSumII.kt
class Solution {
    private val curr: MutableList<Int>
    private val res: MutableList<MutableList<Int>>
    
    init {
        curr = mutableListOf()
        res = mutableListOf()
    }
    
    fun combinationSum2(candidates: IntArray, target: Int): List<List<Int>> {
        candidates.sort()
        aux(candidates, target, 0)
        return res
    }
    
    private fun aux(candidates: IntArray, target: Int, start: Int): Unit {
        if (target == 0) {
            res.add(ArrayList(curr))
            return
        }
        
        if (target > 0) {
            for (i in start..candidates.size-1) {
                if (i > start && candidates[i] == candidates[i-1]) continue
                
                curr.add(candidates[i])
                aux(candidates, target-candidates[i], i+1)
                curr.removeAt(curr.size - 1)
            }
        }
    }
}

File: Problems/Algorithms/40. Combination Sum II/combination_sum_II.rs
impl Solution {
    pub fn combination_sum2(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut results: Vec<Vec<i32>> = Vec::new();
        let mut nums = vec![];
        
        let mut candidates = candidates;
        candidates.sort();
        
        Self::aux(&candidates, &mut results, &mut nums, target, 0);
        
        results
    }
    
    pub fn aux(candidates: &Vec<i32>, results: &mut Vec<Vec<i32>>, nums: &mut Vec<i32>, target: i32, start: usize) {
        if target == 0 {
            results.push(nums.clone());
            return
        }
        
        if target > 0 {
            for i in start..candidates.len() {
                if i > start && candidates[i-1] == candidates[i] {
                    continue
                }
                
                nums.push(candidates[i]);
                Self::aux(candidates, results, nums, target-candidates[i], i+1);
                nums.pop();
            }
        }
    }
}

